envisage/tests/test_core_plugin.py

```diff
@@ -187,3 +175,65 @@ def _preferences_default(self):
 
         # Make sure we can get one of the preferences.
         self.assertEqual("42", application.preferences.get("enthought.test.x"))
+
+    # regression test for enthought/envisage#251
+    def test_unregister_service_offer(self):
+
+        class IJunk(Interface):
+            trash = Str()
+
+        class Junk(HasTraits):
+            trash = Str("garbage")
+
+        class PluginA(Plugin):
+            # The Ids of the extension points that this plugin contributes to.
+            SERVICE_OFFERS = "envisage.service_offers"
+
+            service_offers = List(contributes_to=SERVICE_OFFERS)
+
+            def _service_offers_default(self):
+
+                a_service_offer = ServiceOffer(
+                    protocol=IJunk,
+                    factory=self._create_junk_service,
+                )
+
+                return [a_service_offer]
+
+            def _create_junk_service(self):
+                """ Factory method for the 'Junk' service. """
+
+                return Junk()
+
+            @on_trait_change("application:started")
+            def _unregister_junk_service(self):
+                # only 1 service is registered so it has service_id of 1
+                self.application.unregister_service(1)
+
+        application = TestApplication(
+            plugins=[CorePlugin(), PluginA()],
+        )
+
+        # Run it!
+        return application.run()
+
+    def test_unregister_service(self):
```

Comment:
This test never actually failed with the new fix, but while running I saw the following traceback:
```
Exception occurred in traits notification handler for event object: TraitChangeEvent(object=<envisage.service_registry.ServiceRegistry object at 0x7f993b991e60>, name='unregistered', old=<undefined>, new=1)
Traceback (most recent call last):
  File "/Users/aayres/.edm/envs/envisage-test-3.6-pyqt5/lib/python3.6/site-packages/traits/observation/_trait_event_notifier.py", line 122, in __call__
    self.dispatcher(handler, event)
  File "/Users/aayres/.edm/envs/envisage-test-3.6-pyqt5/lib/python3.6/site-packages/traits/observation/observe.py", line 26, in dispatch_same
    handler(event)
  File "/Users/aayres/Desktop/envisage/envisage/core_plugin.py", line 202, in unregistered_updated
    self._service_ids.remove(event.new)
  File "/Users/aayres/.edm/envs/envisage-test-3.6-pyqt5/lib/python3.6/site-packages/traits/trait_list_object.py", line 801, in remove
    super().remove(value)
  File "/Users/aayres/.edm/envs/envisage-test-3.6-pyqt5/lib/python3.6/site-packages/traits/trait_list_object.py", line 460, in remove
    super().remove(value)
ValueError: list.remove(x): x not in list
```
This test should probably be written so that "failures" like this do not go unnoticed!

Reply:
Adding something like this in the test case's setup and teardown will allow the test to fail:
```
from traits.api import pop_exception_handler, push_exception_handler
...

    def setUp(self):
        push_exception_handler(reraise_exceptions=True)

    def tearDown(self):
        pop_exception_handler()

```

Reply:
hm I added this, but the test is still passing when I remove the fix.  I see:
```
$ python -m unittest test_core_plugin.py 
....Exception occurred in traits notification handler for event object: TraitChangeEvent(object=<envisage.service_registry.ServiceRegistry object at 0x7fb7fb052d00>, name='unregistered', old=<undefined>, new=1)
Traceback (most recent call last):
  File "/Users/aayres/.edm/envs/envisage-test-3.6-pyqt5/lib/python3.6/site-packages/traits/observation/_trait_event_notifier.py", line 122, in __call__
    self.dispatcher(handler, event)
  File "/Users/aayres/.edm/envs/envisage-test-3.6-pyqt5/lib/python3.6/site-packages/traits/observation/observe.py", line 26, in dispatch_same
    handler(event)
  File "/Users/aayres/Desktop/envisage/envisage/core_plugin.py", line 206, in unregistered_updated
    self._service_ids.remove(event.new)
  File "/Users/aayres/.edm/envs/envisage-test-3.6-pyqt5/lib/python3.6/site-packages/traits/trait_list_object.py", line 801, in remove
    super().remove(value)
  File "/Users/aayres/.edm/envs/envisage-test-3.6-pyqt5/lib/python3.6/site-packages/traits/trait_list_object.py", line 460, in remove
    super().remove(value)
ValueError: list.remove(x): x not in list
..
----------------------------------------------------------------------
Ran 6 tests in 0.033s

OK
```

Reply:
ah it seems it is likely related to https://github.com/enthought/traits/issues/1323?

Reply:
> ah it seems it is likely related to [enthought/traits#1323](https://github.com/enthought/traits/issues/1323)?

Using both as suggested in the issue fixed the problem.  I also am realizing I used observe in `CorePlugin` for the fix, but in the new test I wrote I used `on_trait_change` without even realizing.  Should we always be using observe for any new code at this point? @kitchoi

Reply:
> Should we always be using observe for any new code at this point?

Not sure, as this means the next release of envisage will drop support of Traits 6.0, but given envisage is higher up the stack, this might be okay. We should discuss and decide though. To avoid having this PR tied with an unresolved discussion, I'd probably use `on_trait_change` for now.